cmplxfoil/CMPLXFOIL.py

```diff
@@ -0,0 +1,1120 @@
+        for f in evalFuncs:
+            func = self.curAP.name + "_" + f
+            for dvName in DVs.keys():
+                if dvName in self.possibleAeroDVs and dvName in funcsSens[func]:
+                    funcsSens[func][dvName + "_" + self.curAP.name] = funcsSens[func][dvName]
+                    del funcsSens[func][dvName]
+
+    def computeJacobianVectorProductFwd(self, xDvDot=None, xSDot=None, mode="CS", h=None):
+        """This the main python gateway for producing forward mode jacobian
+        vector products. It is not generally called by the user by
+        rather internally or from another solver. A DVGeo object and a
+        mesh object must both be set for this routine.
+
+        Parameters
+        ----------
+        xDvDot : dict
+            Perturbation on the design variables
+        xSDot : numpy array
+            Perturbation on the surface
+        mode : str ["FD" or "CS"]
+            Specifies how the jacobian vector products will be computed
+        h : float
+            Step sized used when the mode is "FD" or "CS" (must be complex
+            if mode = "CS"), by default 1e-6 for FD and 1e-200j for CS
+
+        Returns
+        -------
+        dict
+            Jacobian vector product of evalFuncs given perturbation
+        """
+        if xDvDot is None and xSDot is None:
+            raise ValueError("xDvDot and xSDot cannot both be None")
+
+        if mode not in ["FD", "CS"]:
+            raise ValueError(f'Jacobian vector product mode "{mode}" invalid. Must be either "FD" or "CS"')
+
+        geoDVs = list(self.DVGeo.getValues().keys())
+        possibleDVs = self.possibleAeroDVs + geoDVs
+        for DV in xDvDot.keys():
+            if DV not in possibleDVs:
+                raise ValueError(f'Perturbed design variable "{DV}" is not valid')
+
+        if h is None:
+            if mode == "FD":
+                h = 1e-6
+            elif mode == "CS":
+                h = 1e-200j
+
+        if mode == "FD":
+            orig_funcs = deepcopy(self.funcs[self.curAP.name])
+
+        # Process the Xs perturbation
+        if xSDot is None:
+            xsdot = np.zeros_like(self.coords0)
+        else:
+            xsdot = xSDot
+
+        # For the geometric xDvDot perturbation we accumulate into the
+        # already existing (and possibly nonzero) xsdot and xvdot
+        geoxDvDot = {k: xDvDot[k] for k in geoDVs if k in xDvDot}
+        if xDvDot is not None and self.DVGeo is not None:
+            xsdot += self.DVGeo.totalSensitivityProd(geoxDvDot, self.curAP.ptSetName, config=self.curAP.name).reshape(
+                xsdot.shape
+            )
+
+        # Perturb the coordinates
+        orig_coords = self.getCoordinates()
+        if mode == "FD":
+            self.setCoordinates(orig_coords + xsdot * h)
+        else:
+            self.setCoordinatesComplex(orig_coords + xsdot * h)
```

Comment:
Is this adding a complex perturbation or is `xsdot * h` still real here?

Reply:
Yup, it's handled either by the user-input step or by the `if` statement near the beginning of the method.

Reply:
Didn't catch that. Looks good to me.